### DISCUSSION TITLE
Manually load schema/Switch schema

### DISCUSSION THREAD

**Discussion by ppesti** [Netflix/dgs-framework#1595] [Q&A]
I am building a **multi-purpose GraphQL service** that can boot in different modes. It can be a book, an author, a course, or an assignment service, based on configuration in the application.yaml. This service is also can also be a subgraph for **Apollo Federation**, thus it uses directives such as @key and @external in the schema definition.

Up to this point, I had multiple schemas and only took advantage of one's functionalities. The problem with that was when I had my Book type defined in book.graphqls then the **authorId: ID! field clashed** with the one defined in author.graphqls, which was authorId: ID! @external. Let me demonstrate it a bit better...

book.graphqls:

type Book @key(fields: "id") @key(fields: "authorId") {
    id: ID!
    **authorId: ID!**
    title: String!
    isbn: String
    pages: Int
    yearOfPublication: Int
}

extend type Query {
    books: [Book]
    book(id: ID!): Book
}

----
author.graphqls:

type Author @key(fields: "id") {
    id: ID!
    name: String!
    gender: String
    nationality: String
}

extend type Book @key(fields: "authorId") {
  **authorId: ID! @external**
  author: Author
}

extend type Query {
    authors: [Author]
    author(id: ID): Author
}

This could be remedied by **manually loading in the corresponding schema** (and only that) at startup, though I haven't figured a way to do that. So I could have a directory containing all 4 schemas and just load the desired one based on the value in application.yaml.

If anyone knows how to achieve this, I would greatly appreciate your help!

**Comment by gracecding** [comment]
@ppesti thank you for bringing this up. We will look into it and get back to you.

**Reply by ppesti** [reply]
Thanks a lot @gracecding, I would really appreciate that!

**Reply by gracecding** [reply]
@ppesti When extending a type, the type needs to have the same keys as the original type. Currently, the Book type in author.graphqls only has one of the two keys. This will hopefully fix the issue, but if not, we may be able to diagnose further if you can provide the exact error or your repo.

**Reply by ppesti** [reply]
Thank you for your efforts and response!

Because I am using Apollo Federation, thus I am extending entities, (meaning I define the Book type in both book.graphqls and author.graphqls) I still have **clashing schema definitions** (on the doubly defined fields) in my multi-tenant application, which I wouldn't have without Federation, as it would mean a type is only defined in a single schema.

Luckily, me and my team figured a very simple solution for loading in schemas based on a condition at startup just recently, which I wanted to share.

Instead of placing all our schema files in the resources/schema directory (where it is automatically snatched), we can put that in a separate directory (for example resources/gql), and only placing a single schema.graphqls file in resources/schema. To ensure that the desired schema is available when starting the application/building the jar, we can **create a Gradle task**, which **copies the content** from the corresponding (via env var) schema file, **to resources/schema/schema.graphqls**. But also note that this way, we should either have multipe jars in the image, or have multiple images of our application.

Make sure to specify **dependsOn**, because the copySchema task needs to precede all other build tasks, which use the schema in some way.
